### PR TITLE
Feature/dp 168 api errors mapped to front-end messages

### PR DESCRIPTION
**Pull request by JBaigGoaco**
Feature/dp 168 api errors mapped to front-end messages

image

image

image

### CODE REVIEW

**Review by glgoaco — APPROVED** [review]

File: Frontend/CO.CDP.OrganisationApp.Tests/Pages/Registration/OrganisationDetailsSummaryModelTest.cs [hunk 1]
```diff
@@ -62,6 +63,216 @@ public async Task OnPost_OnSuccess_RedirectsToOrganisationOverview()
...
+        };
+
+        var aex = new ApiException<OrganisationWebApiClient.ProblemDetails>(
+            "Duplicate organisation",
+            400,
+            "Bad Request",
+            null,
+            problemDetails,
+            null
+        );
```
Comment (jakzal) [inline Frontend/CO.CDP.OrganisationApp.Tests/Pages/Registration/OrganisationDetailsSummaryModelTest.cs:90]: What would make these tests more readable is removing irrelevant details. A common pattern is to use a factory method to create fixtures required by the test.


For example. if the status code and error code are the only details that impact the test case, we could reduce it to:

```csharp
var problemDetails = GivenProblemDetail(
  code: ErrorCodes.ORGANISATION_ALREADY_EXISTS,
  status: 400
);
var aex = GivenApiExceptionWithProblemDetails(problem: problemDetails);
```

The idea is that `Given*` methods have all the arguments optional (populated with defaults). This way, the details that are relevant to the test case will be visible in it as they'll be passed to the `Given*` method. Irrelevant details will still be populated with defaults and won't impact test readability.

You can see an example of this approach in the `CO.CDP.OrganisationInformation.Persistence` project, where factory methods are extracted to be reused. However, in your case, I'd just define them as private method inside the test case class until they're needed by another test case.

https://github.com/cabinetoffice/GCGS-Central-Digital-Platform/blob/3735d0958d71579a8736e9cf3d7444c5713a84fb/Services/CO.CDP.OrganisationInformation.Persistence.Tests/EntityFactory.cs#L32-L56

https://github.com/cabinetoffice/GCGS-Central-Digital-Platform/blob/3735d0958d71579a8736e9cf3d7444c5713a84fb/Services/CO.CDP.OrganisationInformation.Persistence.Tests/DatabasePersonRepositoryTest.cs#L38-L51

Comment (jakzal) [reply Frontend/CO.CDP.OrganisationApp.Tests/Pages/Registration/OrganisationDetailsSummaryModelTest.cs:90]: Let me also clarify WHY I think it's a good idea to use factories for test fixtures. 

Object construction can get complex as we often need to provide a lot of arguments. This causes a few problems:
* the test become less readable as it mixes details that are relevant to the test with details that are irrelevant (i.e. `code` is relevant as it participates in decision-making, but the `message` is irrelevant as it's ignored).
* there's lots of duplication between tests.
* tests are more prone to cascade changes caused by constructor signature changes.

Comment (jakzal) [reply Frontend/CO.CDP.OrganisationApp.Tests/Pages/Registration/OrganisationDetailsSummaryModelTest.cs:90]: Let me also clarify WHY I think it's a good idea to use factories for test fixtures. 

Object construction can get complex as we often need to provide a lot of arguments. This causes a few problems:
* the test become less readable as it mixes details that are relevant to the test with details that are irrelevant (i.e. `code` is relevant as it participates in decision-making, but the `message` is irrelevant as it's ignored).
* there's lots of duplication between tests.
* tests are more prone to cascade changes caused by constructor signature changes.

File: Frontend/CO.CDP.OrganisationApp.Tests/Pages/Registration/OrganisationDetailsSummaryModelTest.cs [hunk 2]
```diff
@@ -62,6 +63,216 @@ public async Task OnPost_OnSuccess_RedirectsToOrganisationOverview()
...
+            .ThrowsAsync(aex);
+
+        var model = GivenOrganisationDetailModel();
+
+        await model.OnPost();
+
+        Assert.NotNull(model.ModelState);
+        var modelState = model.ModelState[string.Empty];
+        Assert.NotNull(modelState);
+        modelState.Errors.Should().Contain(e => e.ErrorMessage == ErrorMessagesList.DuplicateOgranisationName);
```
Comment (jakzal) [inline Frontend/CO.CDP.OrganisationApp.Tests/Pages/Registration/OrganisationDetailsSummaryModelTest.cs:105]:
```csharp
model.ModelState[string.Empty].As<ModelStateEntry>().Errors
    .Should().Contain(e => e.ErrorMessage == ErrorMessagesList.PayLoadIssueOrNullAurgument);
```

File: Frontend/CO.CDP.OrganisationApp.Tests/Pages/Registration/OrganisationDetailsSummaryModelTest.cs (hunk 1 above)
Comment (JBaigGoaco) [reply Frontend/CO.CDP.OrganisationApp.Tests/Pages/Registration/OrganisationDetailsSummaryModelTest.cs]: I understand and agree that using factory methods for creating test fixtures will improve the readability and maintainability of the tests. I will refactor the existing unit tests and extract the common code into a TestEntityFactory class.

File: Frontend/CO.CDP.OrganisationApp.Tests/Pages/PersonEntityFactory.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            statusCode,
+            "Bad Request",
+            null,
+            problemDetails ?? GivenProblemDetails(statusCode: statusCode),
+            null
+        );
+        return aex;
+    }
+
+    //public static UserDetails DummyUserDetails()
```
Comment (jakzal) [inline Frontend/CO.CDP.OrganisationApp.Tests/Pages/PersonEntityFactory.cs:44]: Remove comented out code please.

File: Frontend/CO.CDP.OrganisationApp.Tests/Pages/PersonEntityFactory.cs [hunk 4]
```diff
@@ -0,0 +1,66 @@
+using CO.CDP.Person.WebApiClient;
+
+namespace CO.CDP.OrganisationApp.Tests.Pages;
+public static class PersonEntityFactory
```
Comment (jakzal) [inline Frontend/CO.CDP.OrganisationApp.Tests/Pages/PersonEntityFactory.cs:4]: Probably needs a different name. Methods have not much to do with the Person. `ProblemDetailsFactory`? `ErrorFactory`?

File: Frontend/CO.CDP.OrganisationApp.Tests/Pages/Registration/OrganisationDetailsSummaryModelTest.cs [hunk 5]
```diff
@@ -62,6 +63,105 @@ public async Task OnPost_OnSuccess_RedirectsToOrganisationOverview()
             .Which.PageName.Should().Be("/OrganisationOverview");
     }
 
+    [Fact]
+    public async Task OnPost_DuplicateOrganisationName_AddsModelError()
+    {
+        var problemDetails = OrganisationEntityFactory.GivenProblemDetails(statusCode: 400, code: ErrorCodes.ORGANISATION_ALREADY_EXISTS);
+        var aex = OrganisationEntityFactory.GivenApiException(statusCode: 400, problemDetails: problemDetails);
```
Comment (jakzal) [inline Frontend/CO.CDP.OrganisationApp.Tests/Pages/Registration/OrganisationDetailsSummaryModelTest.cs:70]: If you imported the namespaces at the top with:

```csharp
using static CO.CDP.Organisation.WebApiClient.OrganisationEntityFactory;
```

this call could be reduced to a more expressive:


```csharp
 var aex = GivenApiException(statusCode: 400, problemDetails: problemDetails);
```

Comment (jakzal) [inline Frontend/CO.CDP.OrganisationApp.Tests/Pages/Registration/OrganisationDetailsSummaryModelTest.cs:70]: If you imported the namespaces at the top with:

```csharp
using static CO.CDP.Organisation.WebApiClient.OrganisationEntityFactory;
```

this kind of calls could be reduced to a more expressive:


```csharp
 var aex = GivenApiException(statusCode: 400, problemDetails: problemDetails);
```

File: Frontend/CO.CDP.OrganisationApp.Tests/Pages/Registration/OrganisationEntityFactory.cs [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+namespace CO.CDP.OrganisationApp.Tests.Pages.Registration;
+public static class OrganisationEntityFactory
+{
+    public static Organisation.WebApiClient.Organisation GivenClientModel()
+    {
+        return new Organisation.WebApiClient.Organisation(null, null, null, Guid.NewGuid(), null, "Test Org", []);
+    }
+
+    public static ProblemDetails GivenProblemDetails(
```
Comment (jakzal) [inline Frontend/CO.CDP.OrganisationApp.Tests/Pages/Registration/OrganisationEntityFactory.cs:11]: `GivenProblemDetails` and `GivenApiException` definitions seem to be duplicated in a couple of places. Any reason for this?

File: Frontend/CO.CDP.OrganisationApp.Tests/Pages/PersonEntityFactory.cs (hunk 4 above)
Comment (JBaigGoaco) [reply Frontend/CO.CDP.OrganisationApp.Tests/Pages/PersonEntityFactory.cs:4]: yes sure. changing it now.

File: Frontend/CO.CDP.OrganisationApp.Tests/Pages/Registration/OrganisationDetailsSummaryModelTest.cs (hunk 5 above)
Comment (JBaigGoaco) [reply Frontend/CO.CDP.OrganisationApp.Tests/Pages/Registration/OrganisationDetailsSummaryModelTest.cs:70]: yes sure, its updated

File: Frontend/CO.CDP.OrganisationApp.Tests/Pages/Registration/OrganisationEntityFactory.cs (hunk 6 above)
Comment (JBaigGoaco) [reply Frontend/CO.CDP.OrganisationApp.Tests/Pages/Registration/OrganisationEntityFactory.cs:11]: I have prefixed with the called and the type of API has been linked to it.

File: Frontend/CO.CDP.OrganisationApp.Tests/Pages/PersonEntityFactory.cs (hunk 3 above)
Comment (JBaigGoaco) [reply Frontend/CO.CDP.OrganisationApp.Tests/Pages/PersonEntityFactory.cs:44]: well spotted. I was testing some code and forgot to remove it.

File: Frontend/CO.CDP.OrganisationApp.Tests/Pages/Registration/OrganisationEntityFactory.cs (hunk 6 above)
Comment (jakzal) [reply Frontend/CO.CDP.OrganisationApp.Tests/Pages/Registration/OrganisationEntityFactory.cs:11]: It was a confusion on my side. I didn't realise each web api client generates its own problem details type.

**Review by jakzal — APPROVED** [review]

**Pull request merged**